include/rocksdb/configurable.h

```diff
@@ -0,0 +1,335 @@
+  // initialization, such as opening log files or setting up cache parameters.
+  // Implementations should be idempotent (e.g. don't re-open the log file or
+  // reconfigure the cache), as there is the potential this method can be called
+  // more than once.
+  //
+  // By default, this method will also prepare all nested (Inner and
+  // OptionType::kConfigurable) objects.
+  //
+  // @param config_options Controls how the object is prepared.  Also contains
+  //      a Logger and Env that can be used to initialize this object.
+  // @return OK If the object was successfully initialized.
+  // @return InvalidArgument If this object could not be successfull
+  // initialized.
+  virtual Status PrepareOptions(const ConfigOptions& config_options);
+
+  // Checks to see if the settings are valid for this object.
+  // This method checks to see if the input DBOptions and ColumnFamilyOptions
+  // are valid for the settings of this object.  For example, an Env might not
+  // support certain mmap modes or a TableFactory might require certain
+  // settings.
+  //
+  // By default, this method will also validate all nested (Inner and
+  // OptionType::kConfigurable) objects.
+  //
+  // @param db_opts The DBOptions to validate
+  // @param cf_opts The ColumnFamilyOptions to validate
+  // @return OK if the options are valid
+  // @return InvalidArgument If the arguments are not valid for the options
+  //       of the current object.
+  virtual Status ValidateOptions(const DBOptions& db_opts,
+                                 const ColumnFamilyOptions& cf_opts) const;
+
+  // Returns true if this object has been initialized via PrepareOptions, false
+  // otherwise. Once an object has been prepared, only mutable options may be
+  // changed.
+  virtual bool IsPrepared() const { return prepared_; }
+
+ protected:
+  // True once the object is prepared.  Once the object is prepared, only
+  // mutable options can be configured.
+  bool prepared_;
+  // If this class is a wrapper (has-a), this method should be
+  // over-written to return the inner configurable (like an EnvWrapper).
+  // This method should NOT recurse, but should instead return the
+  // direct Inner object.
+  virtual Configurable* Inner() const { return nullptr; }
+
+  // Returns the raw pointer for the associated named option.
+  // The name is typically the name of an option registered via the
+  // Classes may override this method to provide further specialization (such as
+  // returning a sub-option)
+  //
+  // The default implemntation looks at the registered options.  If the
+  // input name matches that of a registered option, the pointer registered
+  // with that name is returned.
+  // e.g,, RegisterOptions("X", &my_ptr, ...); GetOptionsPtr("X") returns
+  // "my_ptr"
+  virtual const void* GetOptionsPtr(const std::string& name) const;
+
+  virtual Status ParseStringOptions(const ConfigOptions& config_options,
```

Comment:
Need some comments for `ParseStringOptions`, `ConfigureOptions` and `ParseOption` here as protected members